### DISCUSSION TITLE
Reset previously loaded adapter firmware

### DISCUSSION THREAD

**Discussion by iron-udjin** [pgj/freebsd-wifibox#139] [General]
Hello,

My adapter is: Intel(R) Wireless-AC 9560 160MHz, REV=0x318
OS: stable/14-n269618-47db906375b5

Possibly, the problem I'm going to describe has already been discussed, but I haven't found a solution yet.

If I boot into Windows and then restart to FreeBSD, the `wifibox` kernel woldn't load firmware correctly:
```
[    0.757319] iwlwifi 0000:00:06.0: loaded firmware version 46.7e3e4b69.0 9000-pu-b0-jf-b0-46.ucode op_mode iwlmvm
[    0.789099] iwlwifi 0000:00:06.0: Detected Intel(R) Wireless-AC 9560 160MHz, REV=0x312
[    0.824412] NET: Registered PF_INET6 protocol family
[    6.191972] iwlwifi 0000:00:06.0: Failed to load firmware chunk!
[    6.191986] iwlwifi 0000:00:06.0: iwlwifi transaction failed, dumping registers
[    6.191991] iwlwifi 0000:00:06.0: iwlwifi device config registers:
[    6.194441] iwlwifi 0000:00:06.0: 00000000: a3708086 00100406 02800010 00000010 c0030004 00000000 00000000 00000000
[    6.194450] iwlwifi 0000:00:06.0: 00000020: 00000000 00000000 00000000 00348086 00000000 000000c8 00000000 000000ff
[    6.194456] iwlwifi 0000:00:06.0: 00000040: 00928010 10000ec0 00100c10 00000000 00000000 00000000 00000000 00000000
[    6.194462] iwlwifi 0000:00:06.0: 00000060: 00000000 00080812 00000405 00000000 00000000 00000000 00000000 00000000
[    6.194467] iwlwifi 0000:00:06.0: 00000080: 000f0011 00002000 00003000 00000000 00000000 00000000 00000000 00000000
[    6.194471] iwlwifi 0000:00:06.0: 000000a0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    6.194477] iwlwifi 0000:00:06.0: 000000c0: 00000000 00000000 c823d001 0d000008 00814005 fee00000 00000000 0000002d
[    6.194482] iwlwifi 0000:00:06.0: 000000e0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    6.194485] iwlwifi 0000:00:06.0: Read failed at 0x100
[    6.194491] iwlwifi 0000:00:06.0: Could not load the [0] uCode section
[    6.194506] iwlwifi 0000:00:06.0: Failed to start INIT ucode: -110
[    6.194514] iwlwifi 0000:00:06.0: WRT: Collecting data: ini trigger 13 fired (delay=0ms).
[    6.444013] iwlwifi 0000:00:06.0: Not valid error log pointer 0x00000000 for Init uCode
```
As a result, the `wifibox` cannot start `hostapd` because the network interface isn't available. Even reboot of `wifibox` or FreeBSD doesn't help.

But... if I shut down the system, unplug the power cable, and push the power button, it will reset the firmware previously loaded by Windows. After that `wifibox` kernel loads firmware correctly and everything works fine:
```
[    0.700582] iwlwifi 0000:00:06.0: loaded firmware version 46.7e3e4b69.0 9000-pu-b0-jf-b0-46.ucode op_mode iwlmvm
[    0.731337] iwlwifi 0000:00:06.0: Detected Intel(R) Wireless-AC 9560 160MHz, REV=0x312
[    0.756317] NET: Registered PF_INET6 protocol family
[    0.791102] iwlwifi 0000:00:06.0: base HW address: 18:56:80:cc:ef:48, OTP minor version: 0x4
[    1.163084] iwlwifi 0000:00:06.0: Registered PHC clock: iwlwifi-PTP, with index: 0
```

It seems the `wifibox` kernel or something else cannot correctly reset previously loaded firmware. By the way, the previous 1.4.X version of `wifibox` didn't have such problem. Somehow, it had been resetting the adapter firmware before.

Most likely, this problem isn't related to `wifibox`. But, is there a way how to _workaround_ this non-critical but annoying issue?

**Comment by pgj** [comment]
I am not an expert on programming PCI devices, but my guess is that Windows leaves the card in some incomplete state that makes the work for both FreeBSD and the virtualized Linux instance harder.  A very cold reboot, as you discovered could be the relevant answer to such problems to my current understanding.  How do you shut down the Windows system?

The reason why you could not see this in earlier version may have something to do with the updates that were done to Wifibox: bump in the Linux kernel version, the `iwlwifi` driver, and the respective firmware.  Which version of Linux do you use in Wifibox?  Note that you can switch between 6.6 and 6.10 in the `net/wifibox-alpine` port, maybe it is worth to try this with the other kernel version.  Alternatively, we can experiment with rolling back the firmware to an older version which is known to be working (which was the March 2024 release for `linux-firmware`, I assume).

**Reply by iron-udjin** [reply]
> How do you shut down the Windows system?

It doesn't matter how I shut down Windows: restart or shutdown. The main thing is to unplug the power cord from the computer case to discharge the capacitors. That's the only thing that performs a firmware reset.

>Which version of Linux do you use in Wifibox?

wifibox-1.5.0
wifibox-alpine-20240911 with 6.6.50-0-lts kernel
wifibox-core-0.14.0

>Note that you can switch between 6.6 and 6.10 in the net/wifibox-alpine

I've rebuilt `net/wifibox-alpine` with the `EDGE` kernel. The same problem as before:
```
[    6.163005] iwlwifi 0000:00:06.0: Could not load the [0] uCode section
[    6.163020] iwlwifi 0000:00:06.0: Failed to start INIT ucode: -110
[    6.163152] iwlwifi 0000:00:06.0: WRT: Collecting data: ini trigger 13 fired (delay=0ms).
[    6.413381] iwlwifi 0000:00:06.0: Not valid error log pointer 0x00000000 for Init uCode
```
>Alternatively, we can experiment with rolling back the firmware 

The current firmware version is `20240909`. To which version should I try rollback?

**Reply by pgj** [reply]
> It doesn't matter how I shut down Windows: restart or shutdown. The main thing is to unplug the power cord from the computer case to discharge the capacitors. That's the only thing that performs a firmware reset.

I was just asking because there were comments about a similar issue with (native) Linux itself that indicated the importance of that.  For example, it could matter if you put your machine in "Airplane mode" which turns off all the wireless devices and then shut down the Windows system.  Along these lines, perhaps it could add something to the mix if you upgrade to the latest version of the Windows drivers.

> The current firmware version is `20240909`. To which version should I try rollback?

Since you said that Wifibox 1.4 worked for you and it included version `20240312` of `linux-firmware`, I would recommend trying that version first.  You can apply the patch below to the `net/wifibox-alpine` and have `linux-firmware` rolled back to that version:

```diff
diff --git a/net/wifibox-alpine/Makefile b/net/wifibox-alpine/Makefile
index 53affd8..088a49c 100644
--- a/net/wifibox-alpine/Makefile
+++ b/net/wifibox-alpine/Makefile
@@ -100,7 +100,7 @@ LINUX_LTS_VERSION=  6.6.50-r0
 LINUX_EDGE_VERSION=    6.10.9-r0
 
 _LINUXFW_SITE= https://git.kernel.org/pub/scm/linux/kernel/git/firmware/
-_LINUXFW_TAG=  20240909
+_LINUXFW_TAG=  20240312
 _LINUXFW=      linux-firmware-${_LINUXFW_TAG}
 
 MASTER_SITES+= ${_GITHUB_SITE}/upstream/:upstream \
diff --git a/net/wifibox-alpine/distinfo b/net/wifibox-alpine/distinfo
index 9a20a4c..0b44cb3 100644
--- a/net/wifibox-alpine/distinfo
+++ b/net/wifibox-alpine/distinfo
@@ -1,8 +1,8 @@
-TIMESTAMP = 1726042467
+TIMESTAMP = 1732087837
 SHA256 (wifibox-alpine/alpine-minirootfs-3.20.3-x86_64.tar.gz) = d4e6fd67dcf75e40c451560ac7265166c2b72a0f38ddc9aae756a7de3d1efa0c
 SIZE (wifibox-alpine/alpine-minirootfs-3.20.3-x86_64.tar.gz) = 3490290
-SHA256 (wifibox-alpine/linux-firmware-20240909.tar.gz) = 93e9b6ae2240661639c874f5fc38f677d18afe365b17a13fee6b4fc4fba42c10
-SIZE (wifibox-alpine/linux-firmware-20240909.tar.gz) = 587006796
+SHA256 (wifibox-alpine/linux-firmware-20240312.tar.gz) = 89cbac35d1bd21ebf64936d764ccd01d4e0b6cde973e3b940f8ad2bac9086ec8
+SIZE (wifibox-alpine/linux-firmware-20240312.tar.gz) = 563675974
 SHA256 (wifibox-alpine/baselayout-3.6.5-r0.apk) = 28f31f534a6fd026d05f3c262dd6519b1b17a528f13a72a1836cc04e062004a7
 SIZE (wifibox-alpine/baselayout-3.6.5-r0.apk) = 13697
 SHA256 (wifibox-alpine/busybox-1.36.1-r4.apk) = 38a40048603eb6f5eacf6d19733101276bfa2afca6976709c089fc7b855cfb0c
```

**Reply by iron-udjin** [reply]
The same problem with `linux-firmware-20240312`:
```
[    0.691360] iwlwifi 0000:00:06.0: loaded firmware version 46.7e3e4b69.0 9000-pu-b0-jf-b0-46.ucode op_mode iwlmvm
[    0.723031] iwlwifi 0000:00:06.0: Detected Intel(R) Wireless-AC 9560 160MHz, REV=0x312
[    0.745667] NET: Registered PF_INET6 protocol family
[    6.168544] iwlwifi 0000:00:06.0: Failed to load firmware chunk!
[    6.168558] iwlwifi 0000:00:06.0: iwlwifi transaction failed, dumping registers
[    6.168564] iwlwifi 0000:00:06.0: iwlwifi device config registers:
[    6.171018] iwlwifi 0000:00:06.0: 00000000: a3708086 00100406 02800010 00000010 c0030004 00000000 00000000 00000000
[    6.171026] iwlwifi 0000:00:06.0: 00000020: 00000000 00000000 00000000 00348086 00000000 000000c8 00000000 000000ff
[    6.171033] iwlwifi 0000:00:06.0: 00000040: 00928010 10000ec0 00100c10 00000000 00000000 00000000 00000000 00000000
[    6.171038] iwlwifi 0000:00:06.0: 00000060: 00000000 00080812 00000405 00000000 00000000 00000000 00000000 00000000
[    6.171044] iwlwifi 0000:00:06.0: 00000080: 000f0011 00002000 00003000 00000000 00000000 00000000 00000000 00000000
[    6.171048] iwlwifi 0000:00:06.0: 000000a0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    6.171054] iwlwifi 0000:00:06.0: 000000c0: 00000000 00000000 c823d001 0d000008 00814005 fee00000 00000000 0000002d
[    6.171059] iwlwifi 0000:00:06.0: 000000e0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    6.171074] iwlwifi 0000:00:06.0: Read failed at 0x100
[    6.171200] iwlwifi 0000:00:06.0: Could not load the [0] uCode section
[    6.171215] iwlwifi 0000:00:06.0: Failed to start INIT ucode: -110
```
`firmware version` is the same as before. Possibly I need to test the older version.

**Reply by pgj** [reply]
Sure.  Note that you can always check out Wifibox 1.4.2 from the developer repository:

https://github.com/pgj/freebsd-wifibox-port/tree/1.4.2

**Reply by pgj** [reply]
@iron-udjin did you have a chance to try 1.4.2 to see this problem shows up?

**Reply by iron-udjin** [reply]
@pgj Hello,

Yes, I tried. The issue still exists. I guess the problem is somewhere in the host OS. Still investigating...
